pkg/sql/opt/exec/execbuilder/tests/local/local_test: TestExecBuild_explain_analyze failed

pkg/sql/opt/exec/execbuilder/tests/local/local_test.TestExecBuild_explain_analyze [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9306209?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9306209?buildTab=artifacts#/) on master @ [bb5c97b5d2073a27e7ec88654b05ac913112f610](https://github.com/cockroachdb/cockroach/commits/bb5c97b5d2073a27e7ec88654b05ac913112f610):


```
=== RUN   TestExecBuild_explain_analyze
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/ec6a1341baa02e55a69448c66fcc4415/logTestExecBuild_explain_analyze237756761
    test_log_scope.go:79: use -show-logs to present logs inline
    logic.go:2841: 
         
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/9282/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/explain_analyze:127: EXPLAIN (VERBOSE) SELECT * FROM system.privileges WHERE path = 'vtable/crdb_internal/tables'
        expected:
            distribution: local
            vectorized: true
            ·
            • scan
              columns: (username, path, privileges, grant_options, user_id)
              estimated row count: 10 (missing stats)
              table: privileges@privileges_path_user_id_key
              spans: /"vtable/crdb_internal/tables"-/"vtable/crdb_internal/tables"/PrefixEnd
        but found (query options: "") :
            distribution: local
            vectorized: true
            ·
            • scan
              columns: (username, path, privileges, grant_options, user_id)
              estimated row count: 1 (100% of the table; stats collected <hidden> ago)
              table: privileges@privileges_path_user_id_key
              spans: /"vtable/crdb_internal/tables"-/"vtable/crdb_internal/tables"/PrefixEnd
[04:50:35] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/9282/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/explain_analyze with config local: 8 tests, 1 failures
[04:50:36] --- total progress: 8 statements
--- total: 8 tests, 1 failures
    logic.go:4121: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/ec6a1341baa02e55a69448c66fcc4415/logTestExecBuild_explain_analyze237756761
--- FAIL: TestExecBuild_explain_analyze (3.11s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #99536 pkg/sql/opt/exec/execbuilder/tests/local/local_test: TestExecBuild_explain_analyze failed [C-test-failure O-robot T-sql-queries branch-release-23.1]
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExecBuild_explain_analyze.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-26120

--- a/pkg/sql/logictest/logic.go
+++ b/pkg/sql/logictest/logic.go
@@ -1294,6 +1294,26 @@ func (t *logicTest) newTestServerCluster(bootstrapBinaryPath string, upgradeBina
 func (t *logicTest) newCluster(
 	serverArgs TestServerArgs, clusterOpts []clusterOpt, knobOpts []knobOpt, toa tenantOverrideArgs,
 ) {
+	makeClusterSettings := func(forSystemTenant bool) *cluster.Settings {
+		var st *cluster.Settings
+		if forSystemTenant {
+			// System tenants use the constructor that doesn't initialize the
+			// cluster version (see makeTestConfigFromParams). This is needed
+			// for local-mixed-22.2-23.1 config.
+			st = cluster.MakeClusterSettings()
+		} else {
+			// Regular tenants use the constructor that initializes the cluster
+			// version (see TestServer.StartTenant).
+			st = cluster.MakeTestingClusterSettings()
+		}
+		// Disable stats collection on system tables before the cluster is
+		// started, otherwise there is a race condition where stats may be
+		// collected before we can disable them with `SET CLUSTER SETTING`. We
+		// disable stats collection on system tables in order to have
+		// deterministic tests.
+		stats.AutomaticStatisticsOnSystemTables.Override(context.Background(), &st.SV, false)
+		return st
+	}
 	var corpusCollectionCallback func(p scplan.Plan, stageIdx int) error
 	if serverArgs.DeclarativeCorpusCollection && t.declarativeCorpusCollector != nil {
 		corpusCollectionCallback = t.declarativeCorpusCollector.GetBeforeStage(t.rootT.Name(), t.t())
@@ -1326,9 +1346,6 @@ func (t *logicTest) newCluster(
 	params := base.TestClusterArgs{
 		ServerArgs: base.TestServerArgs{
 			SQLMemoryPoolSize: serverArgs.MaxSQLMemoryLimit,
-			TempStorageConfig: base.DefaultTestTempStorageConfigWithSize(
-				cluster.MakeTestingClusterSettings(), tempStorageDiskLimit,
-			),
 			DefaultTestTenant: defaultTestTenant,
 			Knobs: base.TestingKnobs{
 				Store: &kvserver.StoreTestingKnobs{
@@ -1416,6 +1433,10 @@ func (t *logicTest) newCluster(
 		} else {
 			require.Lenf(t.rootT, cfg.Localities, 0, "node %d does not have a locality set", i+1)
 		}
+		nodeParams.Settings = makeClusterSettings(true /* forSystemTenant */)
+		nodeParams.TempStorageConfig = base.DefaultTestTempStorageConfigWithSize(
+			nodeParams.Settings, tempStorageDiskLimit,
+		)
 		paramsPerNode[i] = nodeParams
 	}
 	params.ServerArgsPerNode = paramsPerNode
@@ -1427,14 +1448,6 @@ func (t *logicTest) newCluster(
 	stats.DefaultAsOfTime = 10 * time.Millisecond
 	stats.DefaultRefreshInterval = time.Millisecond
 
-	// Disable stats collection on system tables before the cluster is started,
-	// otherwise there is a race condition where stats may be collected before we
-	// can disable them with `SET CLUSTER SETTING`. We disable stats collection on
-	// system tables in order to have deterministic tests.
-	stats.AutomaticStatisticsOnSystemTables.Override(
-		context.Background(), &params.ServerArgs.TempStorageConfig.Settings.SV, false,
-	)
-
 	t.cluster = serverutils.StartNewTestCluster(t.rootT, cfg.NumNodes, params)
 	if cfg.UseFakeSpanResolver {
 		fakeResolver := physicalplanutils.FakeResolverForTestCluster(t.cluster)
@@ -1445,8 +1458,11 @@ func (t *logicTest) newCluster(
 	if cfg.DefaultTestTenant == base.TestTenantEnabled {
 		t.tenantAddrs = make([]string, cfg.NumNodes)
 		for i := 0; i < cfg.NumNodes; i++ {
+			settings := makeClusterSettings(false /* forSystemTenant */)
+			tempStorageConfig := base.DefaultTestTempStorageConfigWithSize(settings, tempStorageDiskLimit)
 			tenantArgs := base.TestTenantArgs{
 				TenantID: serverutils.TestTenantID(),
+				Settings: settings,
 				TestingKnobs: base.TestingKnobs{
 					SQLExecutor: &sql.ExecutorTestingKnobs{
 						DeterministicExplain:            true,
@@ -1458,7 +1474,7 @@ func (t *logicTest) newCluster(
 					RangeFeed: paramsPerNode[i].Knobs.RangeFeed,
 				},
 				MemoryPoolSize:    params.ServerArgs.SQLMemoryPoolSize,
-				TempStorageConfig: &params.ServerArgs.TempStorageConfig,
+				TempStorageConfig: &tempStorageConfig,
 				Locality:          paramsPerNode[i].Locality,
 				TracingDefault:    params.ServerArgs.TracingDefault,
 				// Give every tenant its own ExternalIO directory.

--- a/pkg/sql/opt/exec/execbuilder/testdata/explain_analyze
+++ b/pkg/sql/opt/exec/execbuilder/testdata/explain_analyze
@@ -124,18 +124,14 @@ regions: <hidden>
 statement ok
 GRANT SELECT ON crdb_internal.tables TO root;
 
-# For some reason, even though we explicitly disable the automatic stats
-# collection for both user and system tables, rarely we still see that the stats
-# on system.privileges were collected, so we filter out a single line from the
-# output to make the test deterministic.
 query T
-SELECT * FROM [EXPLAIN (VERBOSE) SELECT * FROM system.privileges WHERE path = 'vtable/crdb_internal/tables']
-  WHERE info NOT LIKE '%estimated row count%';
+EXPLAIN (VERBOSE) SELECT * FROM system.privileges WHERE path = 'vtable/crdb_internal/tables'
 ----
 distribution: local
 vectorized: true
 ·
 • scan
   columns: (username, path, privileges, grant_options, user_id)
+  estimated row count: 10 (missing stats)
   table: privileges@privileges_path_user_id_key
   spans: /"vtable/crdb_internal/tables"-/"vtable/crdb_internal/tables"/PrefixEnd

